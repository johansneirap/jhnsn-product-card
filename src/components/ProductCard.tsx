import React, { createContext } from 'react';
import useProduct from '../hooks/useProduct';
import styles from '../styles/styles.module.css';
import { ProductContextProps, ProductCardProps } from '../interfaces/productInterfaces';



export const productContext = createContext({} as ProductContextProps);
const { Provider } = productContext;

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues } :ProductCardProps) => {
    
    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });

    return (
        <Provider 
            value={{ 
                counter,
                increaseBy, 
                product, 
                maxCount 
            }}>
            <div 
                className={`${ styles.productCard } ${ className }`}
                style={ style }
            >
                { children({
                    count: counter,
                    isMaxCountReached,
                    maxCount: initialValues?.maxCount,
                    product,
                    increaseBy,
                    reset
                }) }
            </div>
        </Provider>
    )
};
