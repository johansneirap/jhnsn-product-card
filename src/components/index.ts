// import { ProductCard as ProductCardHOC } from './ProductCard';

// import { ProductImage } from './ProductImage';
// import { ProductTitle } from './ProductTitle';
// import { ProductButtonsCounter } from './ProductButtonCounter';

export { ProductCard } from './ProductCard';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';
export { ProductButtonsCounter } from './ProductButtonCounter';

// export const ProductCard = Object.assign( ProductCardHOC, {
//     Title: ProductTitle,
//     Image: ProductImage,
//     Buttons: ProductButtonsCounter,
// });