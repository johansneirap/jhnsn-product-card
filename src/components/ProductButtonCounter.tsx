import React, { CSSProperties, useCallback, useContext } from 'react';
import { productContext } from './ProductCard';
import styles from '../styles/styles.module.css';

interface Props {
    className?: string;
    style?: CSSProperties;
}

export const ProductButtonsCounter = ({ className, style }: Props) => {

    const { counter, increaseBy, maxCount } = useContext(productContext);

    const isMaxReached = useCallback(
        () => !!maxCount && counter === maxCount,
        [counter, maxCount]
    );


    return (
        <div
            className={`${styles.buttonsContainer} ${className}`}
            style={style}>
            <button
                className={styles.buttonMinus}
                onClick={() => increaseBy(-1)}
            > - </button>
            <div className={styles.countLabel}> {counter} </div>
            <button
                className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
                onClick={() => increaseBy(+1)}
            > + </button>
        </div>
    );
};