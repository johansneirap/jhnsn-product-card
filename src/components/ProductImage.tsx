import React, { CSSProperties, useContext } from 'react';
import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';
import { productContext } from './ProductCard';

interface Props {
    img?: string | '';
    className?: string;
    style?: CSSProperties;
}

export const ProductImage = ({ img, className, style }: Props) => {
    const { product } = useContext( productContext );
    let imgToShow: string;

    if (img) {
        imgToShow = img;
    } else if (product.img) {
        imgToShow = product.img;
    } else {
        imgToShow = noImage;
    }

    return (
        <img 
            className={`${styles.productImg} ${className}`} 
            src={ imgToShow } 
            alt="Product image" 
            style={ style }
        />
    );
};
