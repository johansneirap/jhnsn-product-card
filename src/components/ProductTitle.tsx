import React, { CSSProperties, useContext } from 'react';
import { productContext } from './ProductCard';
import styles from '../styles/styles.module.css';

interface Props {
    title?: string;
    className?: string;
    style?: CSSProperties;
}

export const ProductTitle = ({ title, className, style }: Props) => {
    const { product } = useContext( productContext );
    return (
        <span 
            className={`${styles.productDescription} ${className}`}
            style={ style }
        >
            { title || product.title }
        </span>
    );
};