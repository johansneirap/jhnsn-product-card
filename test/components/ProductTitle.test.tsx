import React from 'react';
import renderer from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../../src/components'
import { product1 } from '../data/products';

describe('ProductTitle', () => {
  test('debe mostrar el componente correctamente con el titulo personalizado', () => { 
      const wrapper = renderer.create(
          <ProductTitle title="Custom product" className="custom-class"/>
      );
      expect( wrapper.toJSON() ).toMatchSnapshot();
   });

   test('debe mostrar el componente con el nombre del producto', () => { 
       const wrapper = renderer.create(
           <ProductCard product={ product1 }>
               {
                   () => (
                       <ProductTitle title="Custom product" className="custom-class"/>
                   )
               }
           </ProductCard>
       );
       expect( wrapper.toJSON() ).toMatchSnapshot();
    })
});

