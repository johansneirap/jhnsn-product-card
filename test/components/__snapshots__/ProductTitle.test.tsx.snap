// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription custom-class"
  >
    Custom product
  </span>
</div>
`;

exports[`ProductTitle debe mostrar el componente correctamente con el titulo personalizado 1`] = `
<span
  className="productDescription custom-class"
>
  Custom product
</span>
`;
